includes/cli/class-data-backfill.php

```diff
@@ -0,0 +1,365 @@
+			$user_data = [
+				'user_id'  => $user->ID,
+				'email'    => $user->user_email,
+				'metadata' => [
+					// 'current_page_url' is not saved, can't be backfilled.
+					'registration_method' => empty( $registration_method ) ? 'backfill-script' : $registration_method,
+				],
+			];
+			if ( $live ) {
+				self::process_event_entity( $user_data, strtotime( $user->user_registered ), 'reader_registered' );
+			}
+			if ( $verbose ) {
+				WP_CLI::line( '👉 ' . sprintf( 'User %s (#%d) registered on %s.', $user->user_email, $user->ID, $user->user_registered ) );
+			}
+			if ( self::$progress ) {
+				self::$progress->tick();
+			}
+		}
+	}
+
+	/**
+	 * Process donation_new events.
+	 *
+	 * @param string $start The start date.
+	 * @param string $end The end date.
+	 * @param bool   $live Whether to run in live mode.
+	 * @param bool   $verbose Whether to output verbose information.
+	 */
+	private static function process_donation_new( $start, $end, $live, $verbose ) {
+		$orders = wc_get_orders(
+			[
+				'status'       => 'completed',
+				'date_created' => $start . '...' . $end,
+				'limit'        => -1,
+			]
+		);
+		if ( ! $verbose ) {
+			self::$progress = \WP_CLI\Utils\make_progress_bar( 'Processing orders', count( $orders ) );
+		}
+		foreach ( $orders as $order ) {
+			$order_id = $order->get_id();
+			$order_data = \Newspack\Data_Events\Utils::get_order_data( $order_id );
+
+			$timestamp = strtotime( $order->get_date_completed() );
+			if ( $live ) {
+				self::process_event_entity( $order_data, $timestamp, 'donation_new' );
+			}
+			if ( $verbose ) {
+				WP_CLI::line( '👉 ' . sprintf( 'Order #%d completed on %s.', $order_id, gmdate( 'Y-m-d H:i:s', $timestamp ) ) );
+			}
+			if ( self::$progress ) {
+				self::$progress->tick();
+			}
+		}
+	}
+
+	/**
+	 * Process donation_subscription_cancelled events.
+	 *
+	 * @param string $start The start date.
+	 * @param string $end The end date.
+	 * @param bool   $live Whether to run in live mode.
+	 * @param bool   $verbose Whether to output verbose information.
+	 */
+	private static function process_donation_subscription_cancelled( $start, $end, $live, $verbose ) {
+		$subscriptions = wcs_get_subscriptions(
+			[
+				'subscription_status'    => 'cancelled',
+				'subscriptions_per_page' => -1,
+				'meta_query'             => [ // phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_query
+					'relation' => 'AND',
+					[
+						'key'     => wcs_get_date_meta_key( 'cancelled' ),
+						'compare' => '>=',
+						'value'   => $start,
+					],
+					[
+						'key'     => wcs_get_date_meta_key( 'cancelled' ),
+						'compare' => '<=',
+						'value'   => $end,
+					],
+				],
+			]
+		);
+		if ( ! $verbose ) {
+			self::$progress = \WP_CLI\Utils\make_progress_bar( 'Processing subscriptions', count( $subscriptions ) );
+		}
+		foreach ( $subscriptions as $subscription ) {
+			$subscription_data = \Newspack\Data_Events\Utils::get_subscription_data( $subscription );
```

Comment:
again we need to check if it's a donation subscription here. See my comment on the main plugin PR

What happened to me was that I had a bunch of cancellations for non-donation subscriptions. My backfill data was then full of subscription_cancelled events with empy data (because the check was on the Utils function).

So the check should not be there, and be here instead, so we don't include subscription data at all.

Reply:
Handled in 0153127e99ce5135ec0192fcc46e009a57bb4a84